### DISCUSSION TITLE
Further train the el_core_news_lg model for NER with custom dataset

### DISCUSSION THREAD

**Discussion by m-nlp-q** [explosion/spaCy#6972] [Help: Coding & Implementations] [answered]
I want to further train the already trained `el_core_news_lg` model with a custom Greek dataset for NER. 
As of spaCy v3, the training loop have changed, and I wrote it as described in [here](https://spacy.io/usage/v3#migrating-training-python).

I downloaded the model: `python -m spacy download el_core_news_lg`
and this package `pip install spacy-lookups-data` for an errow with the look-up tables 

and my code is the following:

```
import random
import spacy
from spacy.training import Example
import el_core_news_lg

nlp = el_core_news_lg.load()
with open('data.pkl', 'rb') as f:
    train_data_spaCy = pickle.load(f)

ner = nlp.get_pipe('ner')
I2L_spaCy = [
       'B_LOC',
       'B_ORG',
       'B_PERSON',
       'B_MISC',
       'I_LOC',
       'I_ORG',
       'I_PERSON',
       'I_MISC',
       'L_LOC',
       'L_ORG',
       'L_PERSON',
       'L_MISC',
       'U_LOC',
       'U_ORG',
       'U_PERSON',
       'U_MISC',
       'O'
]
for label in I2L_spaCy:
  ner.add_label(label)

examples = []
for text, annots in train_data_spaCy:    
  examples.append(Example.from_dict(nlp.make_doc(text), annots))
  nlp.initialize(lambda: examples)
for i in range(20):
  random.shuffle(examples)
  for batch in minibatch(examples, size=64):
      nlp.update(examples)
```

and line `nlp.initialize(lambda: examples)` throws this error:
**Can't find model 'corpus/el_vectors'. It doesn't seem to be a Python package or a valid path to a data directory.** 

I believe that my data is in the correct format. Here is a sample:

('Μια πρόταση με οντότητα προσώπου τον Κώστα .', {'entities': [(37, 42, 'B_PERSON')]})

Any ideas?

Thank you in advance.

**Comment by svlandeg** [comment, marked as answer]
The error message is quite confusing here. The reason it's thrown, is because the configuration file of `el_core_news_lg` refers to `corpus/el_vectors`, which was used during training of this model, but which isn't present on your system:

```
[paths]
vectors = "corpus/el_vectors"

[initialize]
vectors = ${paths.vectors}
```

The good news is that you don't need to have this data on your system, because you don't have to initialize the pipeline again. Instead of calling
```
nlp.initialize(lambda: examples)
```
you should call
```
nlp.resume_training()
```
and it should work. Let me know if it doesn't!

**Reply by m-nlp-q** [reply]
@svlandeg Thank you for your quick reply.

I changed `nlp.initialize(lambda: examples)` with `nlp.resume_training()` and it runs.

I am new to the library and now I want to monitor training, insert metrics for validation etc. Is there an example for that?
Also, when I saw the errors about `corpus/el_vectors` I thought about token vectors that were missing. Can you provide a short explanation or a source that explains what happens under the hood?

Thanks again.

**Reply by svlandeg** [reply]
Since v3.0 we really recommend using the training [configuration](https://spacy.io/usage/training#config) system instead of writing the training loop yourself. The configuration system is a powerful mechanism to customize your training runs without having to worry about some of the details of the implementation, like when to call `initialize` and when to call `resume_training`. Either way I think it will help your understand of how things work in the background when you read up on the config system.

When using the config file, the default settings will print out the training loss and dev accuracy for each iteration, but you could also plug in other monitoring tools like [Weights & Biases](https://spacy.io/usage/projects#wandb).

If you do want to write the training loop yourself, you could find some inspiration in the code for [`train_while_improving`](https://github.com/explosion/spaCy/blob/master/spacy/training/loop.py#L132) and the [`evaluate`](https://github.com/explosion/spaCy/blob/master/spacy/training/loop.py#L256) function which calls [`nlp.evaluate`](https://spacy.io/api/language#evaluate) with a "dev" corpus.

**Reply by m-nlp-q** [reply]
I prefer to use the training configuration system except for if it does not provide the functionality I want.

My first question regards the dataset. I saw that spaCy works with binary data, and specifically with DocBin. My dataset is in biluo format. Is this format compatible with the greek models and how do I convert the dataset to DocBin? Is there somewhere an example?

Also, I want to measure the performance of the NER model on F1 having excluded the 'O' label, because it will raise the results due to the large volume of words with this label. I saw [here](https://spacy.io/usage/training#metrics) that training configuration supports F score.

**Reply by svlandeg** [reply]
> how do I convert the dataset to DocBin? 

The easiest way is constructing `Doc` objects that contain the annotations you'd want to predict. In your case - ensure `doc.ents` is set correctly. Then you can serialize the created `Doc` objects into the binary `.spacy` file with `DocBin`. Starting from a BILUO annotation scheme, you could use the utility `biluo_tags_to_spans` for this, see more information [here](https://spacy.io/usage/processing-pipelines#wrapping-models-libraries) and [here](https://spacy.io/api/top-level#biluo_tags_to_spans).

```
    from spacy.tokens import DocBin
    from spacy.training import biluo_tags_to_spans

    text = "I flew to Silicon Valley via London."
    doc = nlp(text)
    biluo_tags = ["O", "O", "O", "B-LOC", "L-LOC", "O", "U-GPE", "O"]
    doc.ents = biluo_tags_to_spans(doc, biluo_tags)
    doc_bin = DocBin(docs=[doc])
    doc_bin.to_disk("./data.spacy")
```

**Comment by svlandeg** [comment]
> I prefer to use the training configuration system except for if it does not provide the functionality I want.

Out of curiosity, could you share what type of functionality you'd want but that isn't covered by the config system?

**Reply by m-nlp-q** [reply]
Basically, I am trying to train the model, I don't know if the functionality that I want is provided through training configuration system. 
What I would do in a custom written script is that in every epoch, I would compute micro and macro validation F1 of the categories and keep the weights of the epoch with the highest scores.

Also, I saw that there is an option about CPU and GPU for the Greek models. The GPU option is for transformers. I followed the link to GitHub about the Greek CPU models, but did not find any information about the model.
